auto_uida.c

```diff
@@ -0,0 +1,11 @@
+#include "uidgid.h"
```

Comment:
These files should be autogenerated from a template I guess, they are just too similar.

Reply:
These are 8+2 files that are self-similar.  I did remove the code in common to uid.c + gid.c, respectively, but it is 3 logical lines repeated 8+2 times.

Do you have a concrete suggestion for the kind of concision you'd like to see here?

I could create a new header file, modeled after gen_allocdefs.h and say something like `INITUID(auto_uid, auto_usersa)`.  Or I could use a mechanism similar to how binm*, config, and datemail are generated.  No C files are generated in this way, currently, as C code generation is done in the manner of auto-str.c or auto-int.c.  But I believe the later here is more what you're thinking.

Reply:
I would generate the .c files, they are too dump to keep around ;)

Reply:
+1 to it being a lot of duplication. Usually that worries me more; this code in particular, we probably won't be touching again very often. But if there's a cheap and easy to reduce the duplication, +1 to that too.

Reply:
Please have a look at commit a937324.  This is a patch over the top of the patch that makes uid/gid lookup dynamic.  It creates an auto-id program that generates these files rather than generating them by hand and checking them in.

If this is what you both had in mind, I'll squash these two commits together.  Please let me know.

Reply:
I would have used a auto_id.h.in file with some placeholders and just used sed on them, IIRC other places in the makefile already do this. That avoids building another binary and combining C sources from multiple single strings.

Reply:
Other places in the Makefile do this on shell scripts and manpages, but not for C files, which are generated exclusively by other C programs.  There is a plausible explanation for this, number formatting in auto-int.c and character encoding in auto-str.c.  Yet it is still true that nowhere are we generating C files from shell scripts and sed.

Reply:
On inspection, this feels like one abstraction too far. Pretty sure @DerDakon's suggestion would work for me, and can't think of a reason it'd be a bad idea.

Reply:
I agree with @schmonz, which is why I presented the patch as I originally did.  the code can be fully computed (decided) and the result presented, for the cost of borderline but factually present repetition.  Removing that repetition is too clever by half--it makes things marginally smaller but introduces outsized cognitive overhead.

Reply:
I had a look in where these things are actually used. Look on everything in showctl and qmail-start: this needs no constructors, it can just have functions doing these lookups. Basically everything but qmail-lspawn does never reuse the uid, so computing it once is just extra code for nothing. So at the end we will not need to split these into _any_ extra code, just throw it all away and have one function that does a lookup for a username string or so.

Reply:
since the constructor attribute is, de-minimus, executing code right before main, it has always been true that this lookup code can be inserted in each main that needs it, eliminating the constructor attribute at the cost of repeating the initialization in all uid/gid using programs.

Let me submit a patch that explores those comments and see if we like the non-attribute using, directly inline at main option--with of course the deferred lookups you've identified for qmail-queue.

Reply:
qmail-queue does all of it's lookups in `receivedfmt`, which is called up to two times from `received_setup`, so I cannot consistently (i.e., in the same manner as the rest of the programs) set the uid just before use, so I have left the lookups in main.

The constructor attribute feature is now excised from the code.